File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-4.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        // zxyzxyz
        //     l
        //       r
        // seen set[z,x,y]

        // init left, res, and set
        var left = 0
        var res = 0
        var seen = Set<Character>()
        var chars = Array(s)

        //interate 
        for right in 0..<chars.count {
            // check if we have seen the letter
            while seen.contains(chars[right]){
                seen.remove(chars[left])
                left += 1
            }

            // insert the right char
            seen.insert(chars[right])
            res = max(res, right - left + 1)
        }

        return res
    }
}

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-4.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # [10,1,5,6,7,1]
        #  l
        #     r
        l, r = 0, 1
        maxProfit = 0
        while r < len(prices):
            if prices[r] > prices[l]:
                maxProfit = max(prices[r] - prices[l], maxProfit)
            else:
                l = r
            r += 1
        return maxProfit